File: src/commands/seed.ts
import * as p from '@clack/prompts';
import chalk from 'chalk';
import { join } from 'node:path';
import { exists } from '../utils/fs.js';
import { seedBeads } from '../utils/seed-beads.js';
import { bdCount } from '../utils/bd.js';

interface SeedOptions {
	force?: boolean;
}

export async function seed(specId: string, options: SeedOptions): Promise<void> {
	const cwd = process.cwd();

	p.intro(chalk.bold('forge seed') + chalk.dim(` — ${specId}`));

	const specDir = join(cwd, '.forge', 'specs', specId);

	// Verify spec.yaml exists
	const specYamlPath = join(specDir, 'spec.yaml');
	if (!(await exists(specYamlPath))) {
		p.cancel(`No spec.yaml found at ${specYamlPath}. Run /ingest ${specId} first.`);
		process.exit(1);
	}

	// Check if beads already exist for this spec
	try {
		const existingCount = await bdCount({ labels: [`spec:${specId}`] }, cwd);
		if (existingCount > 0 && !options.force) {
			const overwrite = await p.confirm({
				message: `${existingCount} beads already exist for ${specId}. Re-create?`,
				initialValue: false,
			});
			if (p.isCancel(overwrite) || !overwrite) {
				p.cancel('Seed cancelled.');
				process.exit(0);
			}
		}
	} catch {
		// bd not initialized or count failed — proceed
	}

	const spinner = p.spinner();
	spinner.start('Creating beads from spec.yaml...');

	try {
		const result = await seedBeads(specDir, specId, cwd);
		spinner.stop('Beads created');

		const lines = [
			`Phases:  ${chalk.cyan(String(result.phases))} epics`,
			`Epics:   ${chalk.cyan(String(result.epics))} epics`,
			`Tasks:   ${chalk.cyan(String(result.tasks))} tasks`,
			`Deps:    ${chalk.cyan(String(result.links))} blocking links`,
		];
		p.note(lines.join('\n'), 'Beads seeded');

		p.log.step('Next:');
		p.log.message(`  Run ${chalk.cyan(`forge run ${specId}`)} to start auto-pilot execution`);
		p.log.message(`  Or ${chalk.cyan('bd ready')} to see what tasks are available`);
	} catch (err) {
		spinner.stop('Seeding failed');
		p.log.error(String(err));
		process.exit(1);
	}

	p.outro(chalk.green('Ready to execute.'));
}

File: cmd/seed.go
package cmd

import (
	"fmt"
	"github.com/samahlstrom/forge-cli/internal/bd"
	"github.com/samahlstrom/forge-cli/internal/seedbeads"
	"github.com/samahlstrom/forge-cli/internal/ui"
	"github.com/samahlstrom/forge-cli/internal/util"
	"os"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

var seedForce bool

func init() {
	seedCmd := &cobra.Command{
		Use:   "seed <spec-id>",
		Short: "Create beads (bd tasks) from an approved spec.yaml decomposition",
		Args:  cobra.ExactArgs(1),
		RunE:  runSeed,
	}
	seedCmd.Flags().BoolVar(&seedForce, "force", false, "Re-create beads even if they already exist")
	rootCmd.AddCommand(seedCmd)
}

func runSeed(cmd *cobra.Command, args []string) error {
	specID := args[0]
	cwd, _ := os.Getwd()

	ui.Intro(ui.Bold("forge seed") + ui.Dim(fmt.Sprintf(" — %s", specID)))

	specDir := filepath.Join(cwd, ".forge", "specs", specID)
	if !util.Exists(filepath.Join(specDir, "spec.yaml")) {
		ui.Cancel(fmt.Sprintf("No spec.yaml found at %s. Run /ingest %s first.", filepath.Join(specDir, "spec.yaml"), specID))
		os.Exit(1)
	}

	// Check existing beads
	if !seedForce {
		count, err := bd.Count([]string{fmt.Sprintf("spec:%s", specID)}, cwd)
		if err == nil && count > 0 {
			overwrite, cancelled := ui.Confirm(fmt.Sprintf("%d beads already exist for %s. Re-create?", count, specID), false)
			if cancelled || !overwrite {
				ui.Cancel("Seed cancelled.")
				os.Exit(0)
			}
		}
	}

	spinner := ui.NewSpinner()
	spinner.Start("Creating beads from spec.yaml...")

	result, err := seedbeads.SeedBeads(specDir, specID, cwd)
	if err != nil {
		spinner.Stop("Seeding failed")
		ui.Log.Error(err.Error())
		os.Exit(1)
	}

	spinner.Stop("Beads created")

	lines := []string{
		fmt.Sprintf("Phases:  %s epics", ui.Cyan(fmt.Sprint(result.Phases))),
		fmt.Sprintf("Epics:   %s epics", ui.Cyan(fmt.Sprint(result.Epics))),
		fmt.Sprintf("Tasks:   %s tasks", ui.Cyan(fmt.Sprint(result.Tasks))),
		fmt.Sprintf("Deps:    %s blocking links", ui.Cyan(fmt.Sprint(result.Links))),
	}
	ui.Note(strings.Join(lines, "\n"), "Beads seeded")

	ui.Log.Step("Next:")
	ui.Log.Message(fmt.Sprintf("  Run %s to start auto-pilot execution", ui.Cyan(fmt.Sprintf("forge run %s", specID))))
	ui.Log.Message(fmt.Sprintf("  Or %s to see what tasks are available", ui.Cyan("bd ready")))

	ui.Outro(ui.Green("Ready to execute."))
	return nil
}
